[OC-28] Cli init allows for --templateType to be npm module

nickbalestra opened the pull request:
If the templatetype passed to the init comand is not 'jade' or 'handlebars' (default) but a valid oc-template module, the cli will download it and use it to generate boilerplate.

[Example of boilerplate](https://github.com/opentable/oc-template-react/tree/wip/component) code used by the cli to generate a component

- [x] Coded to easily add yarn in the future 😍  (not yet supported)
- [x] Don't install template if already available on local dev registry
- [x] Use boilerplate code hosted template
- [x] add readme into boilerplates
- [x] logger dependcy injection
- [x] unit tests

A couple of extra notes:
For the moment templates defines "compileDependencies" in their package.json. Those are required and will be installed in order to compile template-specific components and therefore templates to work rely on being initialized through the `oc init` command. The rationale was to avoid having to install all such dependencies on the registry or in places where are not needed at all. Down-size is that for a component creator point of view the api is not too ergonomic as it won't be possible to simply create a component with the right template type and dependencies to work without running `oc init` or manually installing `compileDependencies`. To address this in the near future we might split templates in two modules, a `client` modules and a `compile` module.

matteofigus commented:
I'm not sure what's happening with appveyor. When I click details, it goes to a different PR

matteofigus commented on src/cli/domain/local.js:
```diff
@@ -32,36 +33,39 @@ module.exports = function(){
-        const pathDir = '../../components/base-component-' + templateType,
-          baseComponentDir = path.resolve(__dirname, pathDir),
-          npmIgnorePath = path.resolve(__dirname, pathDir + '/.npmignore');
+      // LEGACY TEMPLATES
```
Can we move this to own file like `init-legacy-templates.js` so that we can make this comment redundant?

matteofigus commented on src/cli/facade/init.js:
```diff
@@ -19,7 +19,7 @@ module.exports = function(dependencies){
     callback = wrapCliCallback(callback);
 
-    local.init(componentName, templateType, (err) => {
+    local.init(componentName, templateType, { logger: logger }, err => {
```
{ logger } ?

matteofigus commented on test/unit/cli-domain-init-template--createComponentDir.js:
```diff
@@ -0,0 +1,53 @@
+    });
+    it('should correctly invoke ensureDirSync', () => {
+      expect(deps['fs-extra'].ensureDirSync.calledWith(
+        config.componentPath
```
personal preference, I like repetition on tests to improve readability. I prefer to read that's called with `path/to/component` over the config, so that if I scroll a lot in the bottom, each tests still tells me what this is doing by  looking at it instead of scrolling back up?

nickbalestra replied:
im fine either way

matteofigus commented on test/unit/cli-domain-init-template--initPackage.js:
```diff
@@ -0,0 +1,33 @@
+  }
+};
+
+const initPackage = injectr('../../src/cli/domain/init-template/initPackage.js', deps, {});
```
This is a minor one, but I think we usually follow conventions of `init-package` for filenames. Are you able to bring all the new files in this PR back to follow the same convention?

matteofigus commented on test/unit/cli-domain-init-template--utils.js:
```diff
@@ -0,0 +1,23 @@
+    it('should return the unaltered package name if alrady given', () => {
+      expect(getPackageName('oc-template-jade')).to.equal('oc-template-jade');
+      expect(getPackageName('are-we-there-yet-1.1.2.tgz')).to.equal('are-we-there-yet');
+      expect(getPackageName('oc-template-jade@next')).to.equal('oc-template-jade');
```
well done on testing all of these scenarios

nickbalestra replied:
😎

nickbalestra commented:
@matteofigus I've change pr base so that now this will merge to master (as the `templates` branch was already been merged). Addressed all the previous points, rebased and improved it a bit further.

matteofigus commented on src/cli/domain/init-template/index.js:
```diff
@@ -0,0 +1,54 @@
+    initPackage({ cli, componentPath });
+    return installTemplate(
+      {
+        componentName,
```
can we make those alphabetic

matteofigus commented on src/cli/domain/init-template/index.js:
```diff
@@ -0,0 +1,54 @@
+      componentPath,
+      compilerPath,
+      logger,
+      callback
```
I would prefer callback not to be in the options? What do you think

matteofigus commented on src/cli/domain/init-template/index.js:
```diff
@@ -0,0 +1,54 @@
+  } = options;
+  const componentPath = path.join(process.cwd(), componentName);
+  const local = /^\.+\/|^\//.test(templateType);
+  const template = utils.getPackageName(templateType).replace('-compiler', '');
```
why are we doing this? If we want to normalise the name, I would rather add this or the next (compiler) into a sanitise method or something like that, in order to make the reader understand a bit more why we are doing this.

matteofigus commented on src/cli/domain/init-template/index.js:
```diff
@@ -0,0 +1,54 @@
+  createComponentDir({ logger, componentName, componentPath });
+
+  try {
+    // If the required compiler is already installed locally no need to install it from NPM again
```
I don't like comments. I would prefer something like
```js
const isInstalledLocally = ...;
const shouldInstallFromNpm = !isInstalledLocally ...
```
What do you think?

nickbalestra replied:
I think I could probably just drop the the comments as it's pretty self understandable imho. 💯  on avoiding comments!

matteofigus commented on src/cli/domain/init-template/install-template.js:
```diff
@@ -0,0 +1,62 @@
+
+  const installProc = spawn(cli, args[cli], { cwd: componentPath });
+
+  installProc.on('error', () => callback('template type not valid'));
```
should we `installing.stop(true)` in case of error?
Also, can we move this messages to `resources`? I think usually the previous convention was to return either a message from `resources` or a code like `TEMPLATE_NOT_VALID` so that later you print the `resources.errors[error_code]....` in a lower level like the facade.

matteofigus commented on src/cli/domain/init-template/install-template.js:
```diff
@@ -0,0 +1,62 @@
+  installProc.on('error', () => callback('template type not valid'));
+  installProc.on('close', code => {
+    if (code !== 0) {
+      return callback('template type not valid');
```
same as before ^

matteofigus commented on src/cli/domain/init-template/scaffold.js:
```diff
@@ -0,0 +1,45 @@
+
+    fs.copySync(baseComponentFiles, componentPath);
+
+    const componentPackage = require(componentPath + '/package.json');
```
I would use readJsonSync as you are requiring something and then overriding its own cache by changing the file immediately after...

matteofigus commented on src/cli/domain/init-template/scaffold.js:
```diff
@@ -0,0 +1,45 @@
+    });
+
+    scaffold.stop();
+    // logger.log(strings.messages.cli.scaffoldSuccess(componentPath));
```
what is this comment?

matteofigus commented on src/cli/domain/local.js:
```diff
@@ -34,34 +36,30 @@ module.exports = function() {
-        fs.outputJsonSync(componentPath, component, { spaces: 2 });
-
-        return callback(null, { ok: true });
+        initTemplate({ componentName, templateType, logger }, callback);
```
initTemplate looks asynchronous to me. So this try/catch should be converted to an async callback check

matteofigus commented on src/resources/index.js:
```diff
@@ -135,9 +138,24 @@ module.exports = {
   },
   messages: {
     cli: {
+      installCompiler: (compiler, fromLocal) =>
```
can we keep those alphabetic. Perhaps first all lower case (new format) then all upper case (old format to be converted to new)

matteofigus commented on src/resources/index.js:
```diff
@@ -135,9 +138,24 @@ module.exports = {
+        `${colors.green('✔')} Files created at ${componentPath}`,
+      creatingDir: () => `Creating directory...`,
+      createdDir: dirName =>
+        `${colors.green('✔')} Created directory "${dirName}"`,
```
I am not sure I like the `${colors.green('✔')}` been here. The responsibility of this file should not be to have colors. I would rather create a new method in the logger, like `logger.okV('message')` which does this for you?

nickbalestra replied:
I'll suggest we do this afterward as we rethink/rearchitect the facade part

nickbalestra commented on src/cli/domain/init-template/install-template.js:
```diff
@@ -0,0 +1,62 @@
+    local,
+    template,
+    logger,
+    callback
```
Not sure I follow you here. The signature is indeed `(option, callback)`. `scaffold` is not in the options cause its been passed as callback

nickbalestra commented:
@matteofigus Thanks for the great review/comments.
After our conversation, aside addressing the review, couple of things I would like to refactor before moving forward:
- move all the logging/ui into the facade, I like this idea.
- remove support for path being passed for templates.

nickbalestra commented on src/cli/domain/init-template/index.js:
```diff
@@ -0,0 +1,54 @@
+    logger = console,
+    cli = 'npm'
+  } = options;
+  const componentPath = path.join(process.cwd(), componentName);
```
It actually happen in one place. The other calls to `process-cwd()` are for different matters (local compiler, resolve for path being passed for template). But agree on hoisting it up in the facade. 👍

nickbalestra commented on src/cli/domain/init-template/success.js:
```diff
@@ -0,0 +1,20 @@
+module.exports = (logger, componentName, componentPath) => {
```
Cause multiline is space sensitive, it will look a bit ugly on that file. I've commited it so that you can see. I think is ok for now as we need to rethink that part anyway.

nickbalestra commented on src/cli/domain/init-template/index.js:
```diff
@@ -0,0 +1,54 @@
+const installTemplate = require('./install-template');
+const createComponentDir = require('./create-component-dir');
+const initPackage = require('./init-package');
+const utils = require('./utils');
```
I wish prettier could do that for me 🤣

matteofigus commented on test/unit/cli-domain-init-template--initPackage.js:
```diff
@@ -0,0 +1,36 @@
+  describe('when invoked', () => {
+    const options = {
+      componentPath: 'path/to/component',
+      cli: 'npm'
```
I think we can remove this line

matteofigus commented on test/unit/cli-domain-init-template--installTemplate.js:
```diff
@@ -0,0 +1,50 @@
+  describe('when invoked', () => {
+    const config = {
+      templateType: 'oc-template-jade',
+      cli: 'npm',
```
here too

matteofigus commented on test/unit/cli-domain-init-template--installTemplate.js:
```diff
@@ -0,0 +1,50 @@
+      templateType: 'oc-template-jade',
+      cli: 'npm',
+      componentPath: 'path/to/component',
+      local: false,
```
same

matteofigus commented on src/cli/domain/init-template/index.js:
```diff
@@ -0,0 +1,61 @@
+    componentName,
+    logger,
+    componentPath,
+    compiler
```
alphabetic

matteofigus commented on src/cli/domain/init-template/install-template.js:
```diff
@@ -0,0 +1,46 @@
+    templateType,
+    compiler,
+    componentPath,
+    logger
```
alphabetic

matteofigus commented on src/cli/domain/init-template/scaffold.js:
```diff
@@ -0,0 +1,43 @@
+    componentPath,
+    compiler,
+    compilerPath,
+    logger
```
alphabetic

matteofigus commented on src/cli/facade/init.js:
```diff
@@ -11,30 +12,45 @@ module.exports = function(dependencies) {
+        templateType,
+        logger,
+        componentPath,
+        compiler: `${templateType}-compiler`
```
alphabetic

nickbalestra commented:
@matteofigus addressed the reviewed points. Some context for you when you'll review it again:
- We weren't doing any validation after the `npm install` of the compiler. As it could be installed any npm module during that process I've added a [validation step as soon as installation succeed](https://github.com/opentable/oc/pull/480/files#diff-438df99ce83948b71454daad0ee79bf0R38) .
- As we already had that validator built-in into `require-template.js` I've [took it ou](https://github.com/opentable/oc/pull/480/files#diff-6e55e1767daec4f38baded3c38ff36b3)t for better [reuse](https://github.com/opentable/oc/pull/480/files#diff-fdfd844163fa3df63f467351abaae684R6).

I've QA the following scenarios (both with the compiler already installed on the dev registry or not):
- `oc init myComponent` -> will setup oc-template-handlebars (work both init & dev, therefore publish)
- `oc init myComponent handlebars` -> will setup oc-template-handlebars (work both init & dev, therefore publish)
- `oc init myComponent jade` -> will setup oc-template-jade (work both init & dev, therefore publish)
- `oc init myComponent oc-template-handlebars/jade` -> will setup oc-template-handlebars/jade (work both init & dev, therefore publish)

All seems good now

matteofigus commented:
LGTM. Let's merge and release tomorrow first thing 👍
